mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractThreadIdGuessingAlgorithmTest.java

```diff
@@ -0,0 +1,186 @@
+
+        // add mails related to old message by subject and In-Reply-To
+        ThreadId threadId4 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.of(new MimeMessageId("Message-ID")), Optional.empty(), Optional.of(new Subject("Re: Test")), mailboxSession).block();
+        ThreadId threadId5 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.of(new MimeMessageId("someInReplyTo")), Optional.empty(), Optional.of(new Subject("Re: Test")), mailboxSession).block();
+        ThreadId threadId6 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.of(new MimeMessageId("references2")), Optional.empty(), Optional.of(new Subject("Fwd: Re: Test")), mailboxSession).block();
+
+        // add mails related to old message by subject and References
+        ThreadId threadId7 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.empty(), Optional.of(List.of(new MimeMessageId("Message-ID"))), Optional.of(new Subject("Fwd: Re: Test")), mailboxSession).block();
+        ThreadId threadId8 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.empty(), Optional.of(List.of(new MimeMessageId("someInReplyTo"))), Optional.of(new Subject("Fwd: Re: Test")), mailboxSession).block();
+        ThreadId threadId9 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.empty(), Optional.of(List.of(new MimeMessageId("references1"), new MimeMessageId("NonRelated-references2"))), Optional.of(new Subject("Fwd: Re: Test")), mailboxSession).block();
+        ThreadId threadId10 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.empty(), Optional.of(List.of(new MimeMessageId("NonRelated-references1"), new MimeMessageId("references2"))), Optional.of(new Subject("Fwd: Re: Test")), mailboxSession).block();
+        ThreadId threadId11 = testee.guessThreadIdReactive(newBasedMessageId, Optional.empty(), Optional.empty(), Optional.of(List.of(new MimeMessageId("references1"), new MimeMessageId("references2"))), Optional.of(new Subject("Fwd: Re: Test")), mailboxSession).block();
+
+        // guess ThreadId should return the same ThreadId with old message
+        assertThat(threadId1).isEqualTo(message.getThreadId());
+        assertThat(threadId2).isEqualTo(message.getThreadId());
+        assertThat(threadId3).isEqualTo(message.getThreadId());
+        assertThat(threadId4).isEqualTo(message.getThreadId());
+        assertThat(threadId5).isEqualTo(message.getThreadId());
+        assertThat(threadId6).isEqualTo(message.getThreadId());
+        assertThat(threadId7).isEqualTo(message.getThreadId());
+        assertThat(threadId8).isEqualTo(message.getThreadId());
+        assertThat(threadId9).isEqualTo(message.getThreadId());
+        assertThat(threadId10).isEqualTo(message.getThreadId());
+        assertThat(threadId11).isEqualTo(message.getThreadId());
```

Comment:
Maybe you could have a look to parametrized tests to be making this more readable?